content/en/blog/_posts/2022-10-31-forensic-container-checkpointing/index.md

```diff
@@ -0,0 +1,204 @@
+---
+layout: blog
+title: "Forensic container checkpointing in Kubernetes"
+date: 2022-10-31
+slug: forensic-container-checkpointing-alpha
+---
+
+**Authors:** Adrian Reber (Red Hat)
+
+Forensic container checkpointing allows the creation of stateful copies of a
+running container without the container knowing that it is being checkpointed.
+The copy of the container can be analyzed and restored in a sandbox environment
+multiple times without the original container being aware of it. Forensic
+container checkpointing was introduced as an alpha feature in Kubernetes v1.25.
+
+## How does it work?
+
+Forensic container checkpointing is based on
+[Checkpoint/Restore In Userspace][criu] (CRIU). With the
+help of CRIU it is possible to checkpoint and restore containers.
+CRIU is integrated in runc, crun, CRI-O and containerd and
+forensic container checkpointing as implemented in Kubernetes
+uses these existing CRIU integrations.
+
+[criu]: https://criu.org/
+
+## Why is it important?
+
+With the help of CRIU and the corresponding integrations it is possible to get
+all information and state about a running container on disk for later forensic
+analysis. Forensic analysis might be important to inspect a suspicious
+container without stopping or influencing it. If the container is really under
+attack, the attacker might detect attempts to inspect the container. Taking a
+checkpoint and analysing the container in a sandboxed environment offers the
+possibility to inspect the container without the original container and maybe
+attacker being aware of the inspection.
+
+In addition to the forensic container checkpointing use case, it is also
+possible to migrate a container from one node to another node without loosing
+the internal state. Especially for stateful containers with long initialization
+times restoring from a checkpoint might save time after a reboot or enable much
+faster startup times.
+
+## How do I use container checkpointing?
+
+The feature is behind a feature gate, so make sure to enable the
+`ContainerCheckpointRestore` gate before you can use the new feature.
```

Comment:
Can we link to some official docs about the feature gate here? Is it a kubelet or API one?

Reply:
It is listed here, but I do not think I can link directly to `ContainerCheckpoint` feature gate description: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

There is also https://kubernetes.io/docs/reference/node/kubelet-checkpoint-api/ . Do you think linking here would make sense?

Reply:
Done